src/SourceTerms/Leakage/CPU_Src_Leakage.cpp

```diff
@@ -0,0 +1,1059 @@
+   const real YeMin      = AuxArray_Flt[SRC_AUX_YEMIN     ];
+   const real YeMax      = AuxArray_Flt[SRC_AUX_YEMAX     ];
+   const real sEint2Code = AuxArray_Flt[SRC_AUX_VSQR2CODE ];
+
+   const int  Mode       = AuxArray_Int[SRC_AUX_MODE      ];
+   const int  NRad_Lin   = AuxArray_Int[SRC_AUX_NRAD_LIN  ];
+   const int  Stride     = AuxArray_Int[SRC_AUX_STRIDE    ];
+   const int  RecMode    = AuxArray_Int[SRC_AUX_RECORD    ];
+
+
+   const int  NRadius = SrcTerms->Leakage_NRadius;
+   const int  NTheta  = SrcTerms->Leakage_NTheta;
+   const int  NPhi    = SrcTerms->Leakage_NPhi;
+#  ifdef __CUDACC__
+   const real *Radius = SrcTerms->Leakage_Radius_DevPtr;
+#  else
+   const real *Radius = h_SrcLeakage_Radius;
+#  endif
+
+   const double x0  = x - PNS_x;
+   const double y0  = y - PNS_y;
+   const double z0  = z - PNS_z;
+   const double rad = sqrt( SQR(x0) + SQR(y0) + SQR(z0) );
+         double theta, phi;
+         int    idx_rad, idx_theta, idx_phi;
+         int    NPhi_half = NPhi>>1;
+         real   xs[2], ys[2], zs[2];
+
+
+// do nothing if the cell is beyond the sampled rays
+   if ( rad > MaxRadius )
+   {
+      if ( RecMode )
+      {
+         fluid[DENS] = (real)0.0;
+         fluid[MOMX] = (real)0.0;
+         fluid[MOMY] = (real)0.0;
+         fluid[MOMZ] = (real)0.0;
+         fluid[ENGY] = (real)0.0;
+#        ifdef YE
+         fluid[YE  ] = (real)0.0;
+#        endif
+      }
+
+#     ifdef DYEDT_NU
+      fluid[DEDT_NU ] = (real)0.0;
+      fluid[DYEDT_NU] = (real)0.0;
+#     endif
+
+      return;
+   }
+
+
+// (1) set up the data index and coordinate
+// (1-1) radius:
+//       --> use data at boundary for cells outside of the radius range
+//       --> for rad <= RadMin_Log, add MIN( ..., NRadius-2 ) to deal the special case RadMin_Log = MaxRadius
+   idx_rad = ( rad <= RadMin_Log )
+           ? MIN(  MAX( int( rad / dRad - (real)0.5 ), 0 ), NRadius-2  )
+           : Src_Leakage_BinarySearch( Radius, NRad_Lin-1, NRadius-2, rad );
+
+   xs[0] = Radius[idx_rad];
+   xs[1] = Radius[idx_rad+1];
+
+// (1-2) theta:
+//       --> for cells close to the pole (theta < 0.5 * dTheta or theta > PI - 0.5 * dTheta)
+//           (a) NPhi > 1, use data at phi + PI for interpolation
+//               --> idx_theta = -1                           if theta <      0.5 * dTheta
+//                               NTheta - 1                   if theta > PI - 0.5 * dTheta
+//                               int( theta / dTheta - 0.5 )  otherwise
+//           (b) NPhi = 1, use data at boundary
+//               --> idx_theta = idx_theta = 0                if theta <      0.5 * dTheta
+//                               NTheta - 2                   if theta > PI - 0.5 * dTheta
+//                               int( theta / dTheta - 0.5 )  otherwise
+   if ( NTheta > 1 )
+   {
+//    set theta = 0.5 * PI arbitrarily if rad = 0
+      theta     = ( rad == (real)0.0 )
+                ? (real)0.5 * M_PI
+                : acos( z0 / rad );
```

Comment:
I am just curious to know. `theta` is predefined as `double`, any reason to have `(real)` here

Reply:
Updated